cmd/thanos/endpointset.go

```diff
@@ -0,0 +1,330 @@
+		queryConnMetricLabels...,
+	)
+}
+
+func setupEndpointsetWithProvider(
+	g *run.Group,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	cfgProvider configProvider,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	// Register resolver for the "thanos:///" scheme for endpoint-groups
+	dns.RegisterGRPCResolver(
+		dns.NewProvider(
+			logger,
+			extprom.WrapRegistererWithPrefix("thanos_query_endpoint_groups_", reg),
+			dns.ResolverType(dnsSDResolver),
+		),
+		dnsSDInterval,
+	)
+	dnsEndpointProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix("thanos_query_endpoints_", reg),
+		dns.ResolverType(dnsSDResolver),
+	)
+	duplicatedStores := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_query_duplicated_store_addresses_total",
+		Help: "The number of times a duplicated store addresses is detected from the different configs in query",
+	})
+	skippedReloading := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_query_skipped_reloading_total",
+		Help: "The number of times we skipped reloading because of errors during endpoint update.",
+	})
+
+	removeDuplicateEndpointSpecs := func(specs []*query.GRPCEndpointSpec) []*query.GRPCEndpointSpec {
+		set := make(map[string]*query.GRPCEndpointSpec)
+		for _, spec := range specs {
+			addr := spec.Addr()
+			if _, ok := set[addr]; ok {
+				level.Warn(logger).Log("msg", "Duplicate store address is provided", "addr", addr)
+				duplicatedStores.Inc()
+			}
+			set[addr] = spec
+		}
+		deduplicated := make([]*query.GRPCEndpointSpec, 0, len(set))
+		for _, value := range set {
+			deduplicated = append(deduplicated, value)
+		}
+		return deduplicated
+	}
+
+	ctxDNSUpdate, cancelDNSUpdate := context.WithCancel(context.Background())
+	g.Add(func() error {
+		return runutil.Repeat(dnsSDInterval, ctxDNSUpdate.Done(), func() error {
+			ctxUpdateIter, cancelUpdateIter := context.WithTimeout(ctxDNSUpdate, dnsSDInterval)
+			defer cancelUpdateIter()
+
+			endpointConfig := cfgProvider.config()
+
+			addresses := make([]string, 0, len(endpointConfig.Endpoints))
+			for _, ecfg := range endpointConfig.Endpoints {
+				if addr := ecfg.Address; dns.IsDynamicNode(addr) {
+					addresses = append(addresses, addr)
+				}
+			}
+			dnsEndpointProvider.Resolve(ctxUpdateIter, addresses)
+			return nil
+		})
+	}, func(error) {
+		cancelDNSUpdate()
+	})
+
+	endpointset := query.NewEndpointSet(time.Now, logger, reg, func() ([]*query.GRPCEndpointSpec, error) {
+		endpointConfig := cfgProvider.config()
+
+		grpcOpts, err := endpointConfig.grpcOpts(logger, reg, nil)
+		if err != nil {
+			skippedReloading.Inc()
+			level.Warn(logger).Log("msg", "Skipping reload of endpoints because of broken gRPC config", "err", err)
+			return nil, err
+		}
+
+		specs := make([]*query.GRPCEndpointSpec, 0)
+		// add static nodes
```

Comment:
Don't we need to call `removeDuplicateEndpointSpecs` on static endpoints too? Can we just call `removeDuplicateEndpointSpecs` on _all_ specs?

Reply:
We didnt before, we can do now of course!